### PR TITLE
fix(builtin): fix regression in 1.6.0 in linker linking root package when under runfiles

### LINKED ISSUE #1823: Regression in 1.6.0 autogenerated test rules
### Affected Rule

Not so sure if it's only autogenerated rules, but the repro I have is failing in a auto generated test rule.

### Is this a regression?
Yes, this works in 1.5.0.

### Description
Auto generated rules fails to load resolve module from a relative file during a test run.

## 🔬 Minimal Reproduction
https://github.com/purkhusid/ts-jest-repro

## 🔥 Exception or Error

<pre><code>
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //:test
-----------------------------------------------------------------------------
[link_node_modules.js] manifest file /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro/_test.module_mappings.json
[link_node_modules.js] manifest contents {
  "workspace": "jest_repro",
  "bin": "bazel-out/darwin-fastbuild/bin",
  "root": "npm/node_modules",
  "modules": {
    "jest_repro": [
      "execroot",
      "bazel-out/darwin-fastbuild/bin"
    ]
  }
}
[link_node_modules.js] startCwd /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro
[link_node_modules.js] isExecroot false
[link_node_modules.js] resolved node_modules root npm/node_modules to /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/npm/node_modules
[link_node_modules.js] cwd /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro
[link_node_modules.js] symlink( node_modules -> /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/npm/node_modules )
[link_node_modules.js] mapping hierarchy [{"name":"jest_repro","link":["execroot","bazel-out/darwin-fastbuild/bin"]}]
[link_node_modules.js] symlink( jest_repro -> /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro/bazel-out/darwin-fastbuild/bin )
bazel node patches enabled. root: /private/var/tmp/_bazel_danielpurkhus/6514ebbf801be819708572b35f1a35af/sandbox/darwin-sandbox/5/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles symlinks in this directory will not escape
FAIL ./lib.test.js
  ● Test suite failed to run

    Cannot find module 'jest_repro/lib' from 'lib.test.js'

      at Resolver.resolveModule (../../../../../node_modules/jest-resolve/build/index.js:259:17)
      at ../../../lib.test.ts:1:1
      at lib.test.js:3:13
      at Object.<anonymous> (lib.test.js:8:3)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        1.085s
Ran all test suites within paths "./lib.test.js".
</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
MacOS 1.14.6
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
Build label: 3.0.0
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Apr 6 12:55:48 2020 (1586177748)
Build timestamp: 1586177748
Build timestamp as int: 1586177748
  </code>
</pre>

**Rules_nodejs version:**
<pre>
  <code>
1.6.0
  </code>
</pre>

### PR DESCRIPTION
**Pull request by gregmagolan**
Resolves #1823 which had a repro https://github.com/purkhusid/ts-jest-repro.

NB: The repo uses ts_library devmode UMD output so the relative require in the source code will be transformed to an absolute require which will include the workspace name.

In 1.5.0:

```
.../sandbox/darwin-sandbox/23/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro/lib.test.js

const lib_1 = require("jest_repro/lib");
```

works as it finds

```
.../sandbox/darwin-sandbox/23/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro/node_modules/jest_repro/lib.js
```

where `test.sh.runfiles/jest_repro/node_modules/jest_repro` is a symlink to the root of the output tree setup by the linker:

```
lrwxr-xr-x  1 greg  wheel  165 29 Apr 05:56 .../sandbox/darwin-sandbox/6/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro/node_modules/jest_repro -> .../sandbox/darwin-sandbox/6/execroot/jest_repro/bazel-out/darwin-fastbuild/bin/test.sh.runfiles/jest_repro
```

In 1.6.0:

Bug introduced in #1850 (thanks for tracking down the culprit @alexeagle) where the root symlink is not created properly when under runfiles (such as a jest_test context).

Fix is:

```
          let runfilesPath = modulePath;
          if (runfilesPath.startsWith(`${bin}/`)) {
            runfilesPath = runfilesPath.slice(bin.length + 1);
+          } else if (runfilesPath === bin) {
+            runfilesPath = '';
          }
```

in `link_node_modules.ts`.

NB: This root workspace symlink will be removed in 2.0 as it causes other unrelated issues. PR for that change is in progress https://github.com/bazelbuild/rules_nodejs/pull/1797. In 2.0, we recommend either using the prodmode esm output of ts_library for jest_test or switching to ts_project where the output flavor is controlled by your tsconfig. An example of using ts_library esm output of ts_library (output_group 'es6_sources') is `under /internal/linker/test/issue_1823_use_ts_library_esm`. It require babel-jest to transform the esm to commonjs. Alternately, you can control the 'es5_sources' module format of ts_library with the `devmode_module` but this can lead to other issues such as the output no longer being compatible with ts_devserver and karma_web_test which require named-UMD or named-AMD.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/rollup/test/integration/far/a/b/c/tsconfig.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
     "strict": true,
     "target": "es2015",
     "module": "esnext",
-    "moduleResolution": "node"
+    "moduleResolution": "node",
+    "types": []
```
Comment (jbedard) [inline packages/rollup/test/integration/far/a/b/c/tsconfig.json:8]: Why add this everywhere?

Comment (gregmagolan) [reply packages/rollup/test/integration/far/a/b/c/tsconfig.json:8]: To keep Windows CI from breaking. Explained in commit message:
This prevents @types packages added to the root package.json from inadvertently breaking tests on buildkite Windows CI where there is no sandboxing.

Comment (gregmagolan) [reply packages/rollup/test/integration/far/a/b/c/tsconfig.json:8]: https://buildkite.com/bazel/rules-nodejs-nodejs/builds/6691#41586486-08b9-447a-aed1-6f9fff24121b

Comment (gregmagolan) [reply packages/rollup/test/integration/far/a/b/c/tsconfig.json:8]:
```

./../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(16,1): error TS6200: Definitions of the following identifiers conflict with those in another file: beforeAll, beforeEach, afterAll, afterEach, describe, fdescribe, xdescribe, it, fit, xit, expect, DEFAULT_TIMEOUT_INTERVAL, CustomMatcherFactory, CustomEqualityTester
--
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(253,15): error TS2428: All declarations of 'ArrayContaining' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(257,15): error TS2428: All declarations of 'ObjectContaining' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(304,9): error TS2687: All declarations of 'message' must have identical modifiers.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(730,15): error TS2428: All declarations of 'SpyAnd' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(747,15): error TS2428: All declarations of 'Calls' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jasmine/ts3.1/index.d.ts(766,15): error TS2428: All declarations of 'CallInfo' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(34,1): error TS6200: Definitions of the following identifiers conflict with those in another file: beforeAll, beforeEach, afterAll, afterEach, describe, fdescribe, xdescribe, it, fit, xit, expect, DEFAULT_TIMEOUT_INTERVAL, CustomMatcherFactory, CustomEqualityTester
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1354,46): error TS2314: Generic type 'ArrayContaining<T>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1355,45): error TS2314: Generic type 'ObjectContaining<T>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1381,15): error TS2428: All declarations of 'ArrayContaining' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1387,15): error TS2428: All declarations of 'ObjectContaining' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1390,9): error TS2386: Overload signatures must all be optional or required.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1396,14): error TS2314: Generic type 'SpyAnd<Fun>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1397,16): error TS2314: Generic type 'Calls<Fun>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1403,15): error TS2428: All declarations of 'SpyAnd' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1436,15): error TS2428: All declarations of 'Calls' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1462,16): error TS2314: Generic type 'CallInfo<Fun>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1467,23): error TS2314: Generic type 'CallInfo<Fun>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1472,18): error TS2314: Generic type 'CallInfo<Fun>' requires 1 type argument(s).
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1479,15): error TS2428: All declarations of 'CallInfo' must have identical type parameters.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1487,9): error TS2717: Subsequent property declarations must have the same type.  Property 'args' must be of type 'Parameters<Fun>', but here has type 'any[]'.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1491,9): error TS2717: Subsequent property declarations must have the same type.  Property 'returnValue' must be of type 'ReturnType<Fun>', but here has type 'any'.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1495,9): error TS2374: Duplicate string index signature.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1515,9): error TS2687: All declarations of 'message' must have identical modifiers.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1515,9): error TS2717: Subsequent property declarations must have the same type.  Property 'message' must be of type 'string \| undefined', but here has type 'string \| (() => string)'.
  | ../../../bk-windows-c4d7/bazel/rules-nodejs-nodejs/node_modules/@types/jest/index.d.ts(1520,9): error TS2375: Duplicate number index signature.
```

Comment (gregmagolan) [reply packages/rollup/test/integration/far/a/b/c/tsconfig.json:8]: Adding @types/jest blew things up and it is generally good practice to be explicit about what types are needed. The other option was `skipLibCheck: true` which is not ideal as it skips checks for all .d.ts files.

**Pull request merged**